qiskit/quantum_info/operators/scalar_op.py

```diff
@@ -0,0 +1,264 @@
+        """Return the composed operator.
+
+        Args:
+            other (BaseOperator): an operator object.
+            qargs (list or None): a list of subsystem positions to apply
+                                  other on. If None apply on all
+                                  subsystems [default: None].
+            front (bool): If True compose using right operator multiplication,
+                          instead of left multiplication [default: False].
+
+        Returns:
+            BaseOperator: The operator self @ other.
+
+        Raises:
+            QiskitError: if other has incompatible dimensions for specified
+                         subsystems.
+
+        Additional Information:
+            Composition (``@``) is defined as `left` matrix multiplication for
+            matrix operators. That is that ``A @ B`` is equal to ``B * A``.
+            Setting ``front=True`` returns `right` matrix multiplication
+            ``A * B`` and is equivalent to the :meth:`dot` method.
+        """
+        if qargs is None:
+            qargs = getattr(other, 'qargs', None)
+
+        if not isinstance(other, BaseOperator):
+            other = Operator(other)
+
+        input_dims, output_dims = self._get_compose_dims(other, qargs, front)
+
+        # If other is also an ScalarOp we only need to possibly
+        # update the coefficient and dimensions
+        if isinstance(other, ScalarOp):
+            if self.coeff is None:
+                coeff = other.coeff
+            elif other.coeff is None:
+                coeff = self.coeff
+            else:
+                coeff = self.coeff * other.coeff
+            return ScalarOp(input_dims, coeff=coeff)
+
+        # If we are composing on the full system we return the
+        # other operator with reshaped dimensions
+        if qargs is None:
+            ret = other.reshape(input_dims, output_dims)
+            if self.coeff is None or self.coeff == 1:
+                return ret
+            return self.coeff * ret
+        # Otherwise compose using other operators method
+        # Note that in this case that operator must know how to initalize
+        # from an ScalarOp either using its to_operator method or having
+        # a specific initialization case.
+        return other.__class__(self).compose(
+            other, qargs=qargs, front=not front)
+
+    def power(self, n):
+        """Return the compose of a operator with itself n times.
+
+        Args:
+            n (int): the number of times to compose with self (n>0).
+
+        Returns:
+            BaseOperator: the n-times composed operator.
+
+        Raises:
+            QiskitError: if the input and output dimensions of the operator
+            are not equal, or the power is not a positive integer.
+        """
+        if self.coeff is None:
+            return self
+        if n == 0:
+            # Raising to zero power returns identity
+            return ScalarOp(self._input_dims)
```

Comment:
even if you remove this block it still does this (only difference is maybe you get `coeff=1+0j` rather than `coeff=1`, which I basically the same). In general I think there are too many manual `if` checking, whereas math will just make things consistent.

Reply:
This was just being pedantic, so could be removed